index.d.ts

```diff
@@ -23,12 +23,18 @@ arrify(undefined);
 //=> []
 ```
 */
-declare function arrify(value: null | undefined): [];
-declare function arrify(value: string): [string];
-declare function arrify<ValueType>(value: ValueType[]): ValueType[];
-// TODO: Use 'readonly ValueType[]' in the next major version
-declare function arrify<ValueType>(value: ReadonlyArray<ValueType>): ReadonlyArray<ValueType>;
-declare function arrify<ValueType>(value: Iterable<ValueType>): ValueType[];
-declare function arrify<ValueType>(value: ValueType): [ValueType];
+declare function arrify<ValueType>(
+	value: ValueType
+): ValueType extends (null | undefined)
+	? []
+	: ValueType extends string
+	? [string]
+	: ValueType extends (infer T)[]
```

Comment:
If you change this to `: ValueType extends ReadonlyArray<unknown> ? ValueType : ...` you also fix #10 
Note that this one condition covers mutable and readonly arrays as well as tuples.